src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1748,50 +1748,74 @@
 
-    (let [transient-memory (mem/to-transient memory)
-          transient-listener (l/to-transient listener)]
-
-      (l/retract-facts! transient-listener facts)
-
-      (binding [*pending-external-retractions* (atom facts)]
-        (external-retract-loop get-alphas-fn transient-memory transport transient-listener))
+    (let [new-pending-operations (conj pending-operations (->PendingUpdate :retraction (if (coll? facts)
+                                                                                         facts
+                                                                                         (into [] facts))))]
 
       (LocalSession. rulebase
-                     (mem/to-persistent! transient-memory)
+                     memory
                      transport
-                     (l/to-persistent! transient-listener)
-                     get-alphas-fn)))
+                     listener
+                     get-alphas-fn
+                     new-pending-operations)))
 
   (fire-rules [session]
 
     (let [transient-memory (mem/to-transient memory)
           transient-listener (l/to-transient listener)]
+
+      ;; We originally performed insertions and retractions immediately after the insert and retract calls,
+      ;; but this had the downside of making a pattern like "Retract facts, insert other facts, and fire the rules"
+      ;; perform at least three transitions between a persistent and transient memory.  Delaying the actual execution
+      ;; of the insertions and retractions until firing the rules allows us to cut this down to a single transition
+      ;; between persistent and transient memory.  There is some cost to the runtime dispatch on operation types here,
+      ;; but this is presumably less significant than the cost of memory transitions.
+      ;;
+      ;; We perform the insertions and retractions in the same order as they were applied to the session since
+      ;; if a fact is not in the session, retracted, and then subsequently inserted it should be in the session at
+      ;; the end.
+      (doseq [{op-type :type facts :facts} pending-operations]
+
+        (condp = op-type
+
+          :insertion
+          (do
+            (l/insert-facts! transient-listener facts)
+
+            (binding [*pending-external-retractions* (atom [])]
+              ;; Bind the external retractions cache so that any logical retractions as a result
+              ;; of these insertions can be cached and executed as a batch instead of eagerly realizing
+              ;; them.  An external insertion of a fact that matches
+              ;; a negation or accumulator condition can cause logical retractions.
+              (doseq [[alpha-roots fact-group] (get-alphas-fn facts)
+                      root alpha-roots]
+                (alpha-activate root fact-group transient-memory transport transient-listener))
+              (external-retract-loop get-alphas-fn transient-memory transport transient-listener)))
```

Comment:
Is there any reason not to do the `*pending-external-retractions*` along with the other given `:retraction`s?

Perhaps it'd just be too tricky to implement for the sake of this PR.